language-support/ts/codegen/tests/ts/daml-json-types/src/index.ts

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+import * as jtv from '@mojotech/json-type-validation';
+
+/**
+ * Interface for companion objects of serializable types. Its main purpose is
+ * to describe the JSON encoding of values of the serializable type.
+ */
+export interface Serializable<T> {
+  // NOTE(MH): This must be a function to allow for mutually recursive decoders.
+  decoder: () => jtv.Decoder<T>;
+}
+
+/**
+ * Identifier of a DAML template.
+ */
+export type TemplateId = {
+  packageId: string;
+  moduleName: string;
+  entityName: string;
+}
+
+/**
+ * Companion object of the `TemplateId` type.
+ */
+const TemplateId: Serializable<TemplateId> = {
+  decoder: () => jtv.object({
+    packageId: jtv.string(),
+    moduleName: jtv.string(),
+    entityName: jtv.string(),
+  })
+}
+
+/**
+ * Interface for objects representing DAML templates. It is similar to the
+ * `Template` type class in DAML.
+ */
+export interface Template<T extends {}> extends Serializable<T> {
+  templateId: TemplateId;
+}
+
+/**
+ * Interface for objects representing DAML choices. It is similar to the
+ * `Choice` type class in DAML.
+ */
+export interface Choice<T, C> extends Serializable<C> {
+  template: Template<T>;
+  choiceName: string;
+}
+
+/**
+ * Interface for objects representing DAML templates with an `Archive` choice.
```

Comment:
Do we really care about supporting templates without an `Archive` choice? Just throwing an error in that case and only having `Template` seems less confusing to users.

Reply:
I'm getting ready for a world where `archive` is a builtin rather than an auto-generated choice. :)

I'd prefer to leave like this for now and maybe even completely remove this in the future.

Reply:
Actually, I'll follow your suggestion. Let's keep it simple for now.